### ISSUE TITLE
Tabs cannot be navigated without using mouse (accessibility)

### ISSUE THREAD

**Issue by slook** [nicotine-plus/nicotine-plus#1485]
Nicotine+ version: 3.1.1  •  GTK 3.24.21
Operating System/Distribution: MX Linux 19 Debian 9 LTS (Stretch) XFCE 4.12

### Describe the bug
When using the interface without a mouse, keyboard trap behavior is experienced when the focus is grabbed by the chat input textbox after activation of the Chat Rooms window tab.

This troublesome text box has a strange autocomplete feature that cycles other users previous wall messages with the `Tab` key, rooms and random commands, and so when the focus becomes trapped onto this textbox it interrupts the expected `Tab` key focus cycling navigation of the main GUI, making it totally impossible to navigate away from this interface object to the Interests tab (or any other tabs shown after the Chat Rooms tab) using only the keyboard right arrow key. Navigation to other interface objects also becomes impossible. No alternative keyboard shortcuts for tab bar navigation are provided.

To make matters worse, the Home key brings the caret to the left of the text box (as expected) instead of the Search Files text box, so it's not even possible for a mouse-less user to return to the start of the GUI focus tab order (the File Search text input box).

Navigation to the Interests window tab is hindered from both sides, because the activating the Search Files tab gives focus away to the search text input box. It then takes considerable effort for the user to cycle back through the focus order so that the Interests window tab can eventually be reached by pressing the left arrow key on the second attempt

It's very difficult, the Chat and Interests features are so inaccessible that the application is basically useless for mouse-less operation.

The only workaround I have found to this chat textbox keyboard trap problem is to press Alt+F4 to exit the application and restart. However, if the "Select previous tab on startup" open is selected, then the interface is totally bricked until a mouse is connected by an assistant (me). I consider any issue that forces the user into an application restart to be a severe bug.

### Expected behavior
When the Chat Rooms window tab is navigated to with a keyboard press (rather than a mouse press), or upon application start, then the chat text input box shouldn't be given the focus, because the user may wish to continue navigating across other main window tabs, such as the Interests tab which is currently inaccessible for a mouse-less user.

When the Search Files window tab is navigated to with a keyboard press (rather than a mouse press), then the search text input box shouldn't be given the focus, because the user may wish to continue navigating across other main window tabs, such as the Interests tab which is currently inaccessible for a mouse-less user.

However, the focus grabbing on mouse press provides a very good usability experience for mouse users to easily enter text into the search and chat textboxes, so this behavior should not be removed, but it is only appropriate for mouse press events. For keyboard users these textboxes need to be first priority so they can be accessible by pressing the `Tab` key once. Then:

- Suggest that if the chat text box was not modified since it gained focus via a keyboard press (rather than a mouse click), then the Tab key should prioritize the GUI navigation tab focus onto other interface objects instead of the chat message autocomplete feature.

- Also suggest, if the string contained in the chat textbox is zero-length, this is a good indicator that autocomplete is not desired, and would allow the user to continue flowing through the other interface objects unhindered by the trap scenario, or once they are done with chatting. The many autocomplete suggestions from a zero-length string are just full of spammy junk anyway!

- Perhaps, the `Tab` key is not the best way to activate the autocomplete feature of the chat message textbox. Suggest `Up` and `Down` keys for this instead, which may also be another way to resolve this keyboard trap bug.

Forcing focus to return to the start of the focus tab order (the file search text box) with `Shift+Home` might also be an idea for a workaround, in addition to the `Home` key without shift (which works in places unless there are search results or another listbox object is focused). Providing a good main window tab navigation solution is ultimately what is needed in an ideal world:

- Ideally, dedicated keyboard shortcuts should be available to enable proper mouse-less operation of the client's GUI main window tabs, this would then provide a method of getting out of any keyboard trap scenario. I suggest `Ctrl+1` thru `Ctrl+9` for main tabs.

- Additionally, dedicated keyboard shortcuts should be available to enable proper mouse-less operation of the Search, Browse and Chat sub-tabs. I suggest `Ctrl+Tab` and `Shift+Ctrl+Tab` for sub tabs.

Whilst providing dedicated keyboard shortcuts for tab bar navigation would be a good workaround for this problem (and provide enhanced usability for all users), I would still consider the keyboard trap issue with the `Tab` key to be a problem because it is a severe accessibility issue that prevents operation of other GUI objects.

I would consider this bug to be resolved if any of the above suggestions can be implemented such that a mouse-less user can navigate away from the chat textbox without restarting the application.

### Steps to reproduce the bug

0. Disconnect mouse (optional step)
1. Press the `Tab` key multiple times to focus the main window tab bar,
2. Use the right arrow key to navigate across to the Interests window tab,
3. The focus is trapped in the chat text input box,
4. The client's interface is bricked.

### Additional context

This is an accessibility problem, see link:
[Web Content Accessibility Guidelines to Non-Web Information and Communications Technologies (WCAG2ICT) - Quick Reference](https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=212#no-keyboard-trap)

Guideline 2.1.2 (No Keyboard Trap) requires that "If keyboard focus can be moved to a component of the page using a keyboard interface, then focus can be moved away from that component using only a keyboard interface, and, if it requires more than unmodified arrow or tab keys or other standard exit methods, the user is advised of the method for moving focus away."

See also link:
[Techniques for WCAG - Ensuring that users are not trapped in content](https://www.w3.org/WAI/WCAG21/Techniques/general/G21)

There could be many reasons why a user may be unable to use a mouse or other pointing device... disability, visual impairment, cramped workspace (such as a DJ console), bright flashing lights at a party and/or being too drunk!

**Comment by mathiascode** [comment]
> But this slightly breaks the autocomplete feature, only one press of the Tab works for the first suggestion only (which is kind of okay), then the focus moves away to the Room Wall button. Ideally, the focus should only be allowed to move away if the textbox is empty.

Fixed
 
> Navigating the focus to the main window tab bar is still troublesome, but at least it's possible.

Looking into the best way to do this. Is there any universal shortcut for this? There's at least F6 which changes focus between the address bar and main content in web browsers.

Our keyboard shortcuts are broken on some systems. Currently investigating.

**Comment by mathiascode** [comment]
Does something like Ctrl+F work in the log pane/chat views? The Gtk-CRITICAL error doesn't seem to be related to the keyboard shortcuts, and the keyboard shortcuts work fine in my MX-Linux 18.3 VM with GTK 3.22.11.

**Comment by slook** [comment]
Does something like Ctrl+F work in the log pane/chat views?

> Yes

Maybe my MX18.3 install is a bit borked, it was installed way back, like years ago, to bare metal for longevity testing of the MX updates and is not intended to be a pristine development environment or testing rig. And I have hacked around with the key bindings because it's a laptop with special function keys for sleep, volume and brightness etc.

Unfortunelty my desktop rig is in pieces at the moment, keep running out of hard drives :-/

I have a MX-Linux 19.4 VM on here tho:

```
setup@VBoxMX:~/Git/nicotine-plus$ ./nicotine
Cannot find gdbm or semidbm, please install either one.

```
brb...

**Comment by slook** [comment]
Found python3-gdbm package that got it working.

Surprisingly, it the same, tab hotkeys not work. This makes me think it is a kernel driver problem with the ThinkPad keyboard, but I do not have such issues with any other applications.

I shall go find an external USB keyboard, brb...

**Comment by slook** [comment]
Nope still the same with external keyboard.

This is strange, I'm thinking it is perhaps a laptop specific quirk with the kernel keyboard driver or something.

Setting the System Default keyboard driver in my VM has no effect, I need to logout and back in to swap the driver on the host.

brb...

**Comment by slook** [comment]
Nope still the same with standard keyb driver.

I can't think of any other GTK3 apps that have a tab bar like N+ ... I suppose it's possible that I wouldn't know even if I had some issue with my other GTK3 apps, although keyboard navigation seems to be alright everywhere else.

There seems to be an edge case. I can't see that your changes would introduce any new bugs that didn't already exist regards key binding conflicts or whatnot. So the failure of my test is no reason not to push your commit as planned.

**Comment by mathiascode** [comment]
I pushed a minor change for the shortcuts (https://github.com/nicotine-plus/nicotine-plus/commit/15cc5d358e2829a63ad7f0ecf9775b3fb1c536ec). Let me know if it helps.

**Comment by slook** [comment]
Yes well done! Amazing the both tab rows working :-)

The autocomplete is also working for multiple suggestions in the chat textbox :-)

Wow it is a massive improvement in usability of the UI

**Comment by slook** [comment]
The main window tab bar works great everytime, no matter what object has the focus, I love it!

The child tabs in the Search Files sections do not work unless one of the tabs on tab bar itself has focus, this happens directly after submitting a new search but the first thing that will be needed is scroll down (dn) or page down (pgdn), so the ListBox needs to have the focus after a new search tab is created.

It would be expected that the child search tabs should switch no matter what currently has the focus, such as the child chat room tabs (which work okay as expected, even from within the textbox which is cool).

I suppose it's also important that the focus remains on the search results listbox so that scrolling (up/down) and paging (PgUp/PgDn) is always possible, but pressing the down arrow key after selecting a different tab goes to the Results counter button and then down the DropDownListBox below that and starts advancing through previous 'Include text..." entries, which is not wanted and creates a navigation obstruction.

- Every time a child search tab is activated (either with mouse, with key or with new search) then the ListBox needs the focus passed to it.

The new shortcut keys are worthy of being documented in the Keyboard Shortcuts window under the 'Tabs' section.

**Comment by mathiascode** [comment]
> Every time a child search tab is activated (either with mouse, with key or with new search) then the ListBox needs the focus passed to it.

Will look into it.

> The Ctrl+Tab and Shift+Tab key bindings need to have a wider scope so child tabs can be switched anytime while the Search Files tab is active.

See the end of https://github.com/nicotine-plus/nicotine-plus/issues/1485#issuecomment-907438357. I'm still trying to figure out how to override the behavior.

**Comment by mathiascode** [comment]
> Every time a child search tab is activated (either with mouse, with key or with new search) then the ListBox needs the focus passed to it.

This is now done.

**Comment by slook** [comment]
> just a minor issue left where GtkTreeView swallows the key input due to conflicting shortcuts

I can see you have made partial progress on this, Ctrl+Tab and Shift+Ctrl+Tab kindof works, but only when the Search tab bar is positioned on the Right side in Preferences, and only when the key is pressed twice, it's kinda wonky but that's cool for a start :-)

Shame it doesn't work in any other positions... If you are going to make the child Search tab bar position static, then I would personally prefer the Left side seems to make the most sense from a UI flow perspective, especially when the Search Files tab is in position 1. (btw, the Chat rooms bar tab position is best on the Right imo).

**Comment by mathiascode** [comment]
> > just a minor issue left where GtkTreeView swallows the key input due to conflicting shortcuts
> 
> I can see you have made partial progress on this, Ctrl+Tab and Shift+Ctrl+Tab kindof works, but only when the Search tab bar is positioned on the Right side in Preferences, and only when the key is pressed twice, it's kinda wonky but that's cool for a start :-)
> 
> Shame it doesn't work in any other positions... If you are going to make the child Search tab bar position static, then I would personally prefer the Left side seems to make the most sense from a UI flow perspective, especially when the Search Files tab is in position 1. (btw, the Chat rooms bar tab position is best on the Right imo).

The problem is that the treeview completely ignores our custom Ctrl+Tab behavior, and continues to move the focus to the next widget like Tab normally does. This is likely why it kind of works when the tab bar is on the right side, since the tab bar widget is next in line after the treeview.

I'll figure out a solution, but it may take some time.

**Comment by mathiascode** [comment]
This should be fixed now, and you should be able to use Ctrl+Tab no matter which widget is focused.

**Comment by slook** [comment]
Honestly @mathiascode , this fix is sensational :-)

> I'll figure out a solution, but it may take some time.

Well I was expecting like at least a few weeks or whatnot, seems like you are really keen to get this whole thing going proper. I admire your dedication.
